src/main/java/hudson/remoting/Engine.java

```diff
@@ -196,11 +203,87 @@ public void removeListener(EngineListener el) {
+                try {
+                    store.load(null, password);
+                } catch (NoSuchAlgorithmException e) {
+                    throw new IllegalStateException("Java runtime specification requires support for JKS key store", e);
+                } catch (CertificateException e) {
+                    throw new IllegalStateException("Empty keystore", e);
+                }
+                KeyManagerFactory kmf;
+                try {
+                    kmf = KeyManagerFactory.getInstance(KeyManagerFactory.getDefaultAlgorithm());
+                } catch (NoSuchAlgorithmException e) {
+                    throw new IllegalStateException("Java runtime specification requires support for default key manager", e);
+                }
+                try {
+                    kmf.init(store, password);
+                } catch (KeyStoreException e) {
+                    throw new IllegalStateException(e);
+                } catch (NoSuchAlgorithmException e) {
+                    throw new IllegalStateException(e);
+                } catch (UnrecoverableKeyException e) {
+                    throw new IllegalStateException(e);
+                }
+                try {
+                    context.init(kmf.getKeyManagers(), new TrustManager[]{agentTrustManager}, null);
+                } catch (KeyManagementException e) {
+                    events.error(e);
+                    return;
+                }
+                innerRun(hub, context, executor);
+            } finally {
+                hub.close();
+            }
+        } catch (IOException e) {
+            events.error(e);
+        }
+    }
+
+    @SuppressWarnings({"ThrowableInstanceNeverThrown"})
+    private void innerRun(IOHub hub, SSLContext context, ExecutorService service) {
         // Create the protocols that will be attempted to connect to the master.
-        List<JnlpProtocol> protocols = JnlpProtocolFactory.createProtocols(secretKey, slaveName, events);
+        List<JnlpProtocolHandler> protocols = new JnlpProtocolHandlerFactory(service)
+                .withIOHub(hub)
+                .withSSLContext(context)
+                .withPreferNonBlockingIO(false) // we only have one connection, prefer blocking I/O
+                .handlers();
+        final Map<String,String> headers = new HashMap<String,String>();
+        headers.put(JnlpConnectionState.CLIENT_NAME_KEY, slaveName);
+        headers.put(JnlpConnectionState.SECRET_KEY, secretKey);
+        List<String> jenkinsUrls = new ArrayList<String>();
+        for (URL url: candidateUrls) {
+            jenkinsUrls.add(url.toExternalForm());
+        }
+        JnlpAgentEndpointResolver resolver = new JnlpAgentEndpointResolver(jenkinsUrls);
+        resolver.setCredentials(credentials);
+        resolver.setProxyCredentials(proxyCredentials);
+        resolver.setTunnel(tunnel);
+        try {
+            resolver.setSslSocketFactory(getSSLSocketFactory());
+        } catch (Exception e) {
```

Comment:
We just throw a dozen of different exceptions from the method in order to handle them in such way